src/awkward/_v2/index.py

```diff
@@ -1,79 +1,84 @@
+        np.dtype(np.uint32): "u32",
+        np.dtype(np.int64): "i64",
+    }
 
-class Index:
     def __init__(self, data):
         self._nplike = ak.nplike.of(data)
+
         self._data = self._nplike.asarray(data, order="C")
+        if len(self._data.shape) != 1:
+            raise TypeError("an actual error string")
+
         self._T = self._data.dtype
-        if self._data == None:
-            self._len = 0
-        else:
-            self._len = len(self.data)
-        if self._T not in (
-            self._nplike.int8,
-            self._nplike.uint8,
-            self._nplike.int32,
-            self._nplike.uint32,
-            self._nplike.int64,
-            object,
-        ):
-            raise TypeError
-
-    # the default for this is float64 so maybe type should also be mentioned
-    def __zeros__(length, nplike):
-        return nplike.zeros(length)
-
-    # the default for this is float64 so maybe type should also be mentioned
-    def __empty__(length, nplike):
-        return nplike.empty(length)
-
-    def __data__(self):
+        if self._T not in self._dtype_to_form:
+            raise TypeError("an actual error string")
+
+    @classmethod
+    def zeros(cls, length, nplike, dtype):
+        return nplike.zeros(length, dtype=dtype)
+
+    @classmethod
+    def empty(cls, length, nplike, dtype):
+        return nplike.empty(length, dtype=dtype)
+
+    @property
+    def data(self):
         return self._data
 
-    def __nplike__(self):
+    @property
+    def nplike(self):
         return self._nplike
 
     def __len__(self):
-        return self._len
+        if self._data is None:
+            return 0
+        else:
+            return len(self._data)
 
-    # array_str needs to be included in nplike
     def __repr__(self):
+        # FIXME
         return self._nplike.array_str(self._data)
 
-    def __form__(self):
-        type = str(self._T)
-        return type[0] + type.split("t")[1]
+    def form(self):
+        return self._dtype_to_form[self._data.dtype]
 
     def __getitem__(self, index):
         return self._data[index]
 
     def __setitem__(self, index, value):
         self._data[index] = value
 
-    # the error that might occur from can be handled by the nplike, but if start stop are not in range it returns an empty array, is this behaviour constant with that it's expected?
-    def __getitem_range__(self, start, stop):
-        if start < stop and stop < self._len:
-            return self._data[start:stop]
-        else:
-            print("Illegal start:stop for this length")
-
-    def __to64__(self):
-        self._data = self._data.astype("int64")
+    def to64(self):
+        return Index(self._data.astype(np.int64))
 
-    def __iscontiguous__(self):
-        return self._data.strides[0] == self._data.itemsize
+    def iscontiguous(self):
+        if self._data is None:
+            return True
+        else:
+            return self._data.strides == (self._data.itemsize,)
```

Comment:
Same here: should return an Index instance. Also handling the case of `self._data is None`. Also, note that comparisons with `None` use the Python `is` rather than `==`. (See [this if you're wondering](https://stackoverflow.com/q/26595/1623645).)

Also also, the expression

```python
self._data.strides == (self._data.itemsize,)
```

is more restrictive than the expression

```python
self._data.strides[0] == self._data.itemsize
```

since the latter allows `self._data.strides` to have more dimensions. We've constrained `self._data.shape` to one dimension in the constructor and presumably that restricts the `self._data.strides` as well, but we might as well be more careful.